File: Wire-iOS/Sources/UserInterface/Components/Views/UpsideDownTableView.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "UpsideDownTableView.h"

@implementation UpsideDownTableView

- (instancetype)initWithFrame:(CGRect)frame style:(UITableViewStyle)style
{
    self = [super initWithFrame:frame style:style];
    
    if (self) {
        [UIView performWithoutAnimation:^{
            self.transform = CGAffineTransformMakeScale(1, -1);
        }];
    }
    
    return self;
}

- (void)setCorrectedContentInset:(UIEdgeInsets)contentInset
{
    [super setContentInset:UIEdgeInsetsMake(contentInset.bottom, contentInset.left, contentInset.top, contentInset.right)];
}

- (UIEdgeInsets)correctedContentInset
{
    UIEdgeInsets insets = [super contentInset];
    return UIEdgeInsetsMake(insets.bottom, insets.left, insets.top, insets.right);
}
    
- (UIEdgeInsets)correctedScrollIndicatorInsets
{
    UIEdgeInsets insets = super.scrollIndicatorInsets;
    return UIEdgeInsetsMake(insets.bottom, insets.left, insets.top, insets.right);
}
    
- (void)setCorrectedScrollIndicatorInsets:(UIEdgeInsets)correctedScrollIndicatorInsets
{
    [super setScrollIndicatorInsets:UIEdgeInsetsMake(correctedScrollIndicatorInsets.bottom, correctedScrollIndicatorInsets.left, correctedScrollIndicatorInsets.top, correctedScrollIndicatorInsets.right)];
}

- (void)setTableHeaderView:(UIView *)tableHeaderView
{
    tableHeaderView.transform = CGAffineTransformMakeScale(1, -1);
    
    [super setTableFooterView:tableHeaderView];
}

- (UIView *)tableHeaderView
{
    return super.tableFooterView;
}

- (void)setTableFooterView:(UIView *)tableFooterView
{
    tableFooterView.transform = CGAffineTransformMakeScale(1, -1);
    
    [super setTableHeaderView:tableFooterView];
}

- (UIView *)tableFooterView
{
    return super.tableHeaderView;
}

- (id)dequeueReusableCellWithIdentifier:(NSString *)identifier
{
    UITableViewCell *cell = [super dequeueReusableCellWithIdentifier:identifier];
    
    cell.transform = CGAffineTransformMakeScale(1, -1);
    
    return cell;
}

- (void)scrollToNearestSelectedRowAtScrollPosition:(UITableViewScrollPosition)scrollPosition animated:(BOOL)animated
{
    [super scrollToNearestSelectedRowAtScrollPosition:[self inverseScrollPosition:scrollPosition] animated:animated];
}

- (void)scrollToRowAtIndexPath:(NSIndexPath *)indexPath atScrollPosition:(UITableViewScrollPosition)scrollPosition animated:(BOOL)animated
{
    [super scrollToRowAtIndexPath:indexPath atScrollPosition:[self inverseScrollPosition:scrollPosition] animated:animated];
}

- (id)dequeueReusableCellWithIdentifier:(NSString *)identifier forIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [super dequeueReusableCellWithIdentifier:identifier forIndexPath:indexPath];
    
    cell.transform = CGAffineTransformMakeScale(1, -1);
    
    return cell;
}

- (UITableViewScrollPosition)inverseScrollPosition:(UITableViewScrollPosition)scrollPosition
{
    if (scrollPosition == UITableViewScrollPositionTop) {
        return UITableViewScrollPositionBottom;
    }
    else if (scrollPosition == UITableViewScrollPositionBottom) {
        return UITableViewScrollPositionTop;
    }
    else {
        return scrollPosition;
    }
}

@end

File: wire-ios/Wire-iOS/Sources/UserInterface/Components/Views/UpsideDownTableView.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit

final class UpsideDownTableView: UITableView {

    /// The view that allow pan gesture to scroll the tableview
    weak var pannableView: UIView?

    override init(frame: CGRect, style: UITableView.Style) {
        super.init(frame: frame, style: style)

        UIView.performWithoutAnimation({
            self.transform = CGAffineTransform(scaleX: 1, y: -1)
        })
    }

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    var correctedContentInset: UIEdgeInsets {
        get {
            let insets = super.contentInset
            return UIEdgeInsets(top: insets.bottom, left: insets.left, bottom: insets.top, right: insets.right)
        }

        set {
            super.contentInset = UIEdgeInsets(top: newValue.bottom,
                                                  left: newValue.left,
                                                  bottom: newValue.top,
                                                  right: newValue.right)
        }
    }

    var lockContentOffset: Bool = false

    override var contentOffset: CGPoint {
        get {
            return super.contentOffset
        }

        set {
            // Blindly ignoring the SOLID principles, we are modifying the functionality of the parent class.
            if lockContentOffset {
                return
            }
            /// do not set contentOffset if the user is panning on the bottom edge of pannableView (with 10 pt threshold)
            if let pannableView,
               self.panGestureRecognizer.location(in: self.superview).y >= pannableView.frame.maxY - 10 {
                return
            }

            super.contentOffset = newValue
        }
    }

    override var tableHeaderView: UIView? {
        get {
            return super.tableFooterView
        }

        set(tableHeaderView) {
            tableHeaderView?.transform = CGAffineTransform(scaleX: 1, y: -1)
            super.tableFooterView = tableHeaderView
        }
    }

    override var tableFooterView: UIView? {
        get {
            return super.tableHeaderView
        }

        set(tableFooterView) {
            tableFooterView?.transform = CGAffineTransform(scaleX: 1, y: -1)
            super.tableHeaderView = tableFooterView
        }
    }

    override func dequeueReusableCell(withIdentifier identifier: String) -> UITableViewCell? {
        let cell = super.dequeueReusableCell(withIdentifier: identifier)
        cell?.transform = CGAffineTransform(scaleX: 1, y: -1)
        return cell
    }

    override func scrollToNearestSelectedRow(at scrollPosition: UITableView.ScrollPosition, animated: Bool) {
        super.scrollToNearestSelectedRow(at: inverseScrollPosition(scrollPosition), animated: animated)
    }

    override func scrollToRow(at indexPath: IndexPath,
                              at scrollPosition: UITableView.ScrollPosition,
                              animated: Bool) {
        super.scrollToRow(at: indexPath, at: inverseScrollPosition(scrollPosition), animated: animated)
    }

    override func dequeueReusableCell(withIdentifier identifier: String, for indexPath: IndexPath) -> UITableViewCell {
        let cell = super.dequeueReusableCell(withIdentifier: identifier, for: indexPath)

        cell.transform = CGAffineTransform(scaleX: 1, y: -1)

        return cell
    }

    func inverseScrollPosition(_ scrollPosition: UITableView.ScrollPosition) -> UITableView.ScrollPosition {
        if scrollPosition == .top {
            return .bottom
        } else if scrollPosition == .bottom {
            return .top
        } else {
            return scrollPosition
        }
    }
}
